libr/include/r_vec.h

```diff
@@ -0,0 +1,346 @@
+ *
+ * Note that compared to the older r_vector and r_pvector types in R2, there is
+ * only a single type that exists for both usecases.
+ *
+ * The way to use this code is as follows:
+ * 1. Use the R_GENERATE_VEC_IMPL_FOR macro to generate the vector implementation for a certain type
+ * 2. Use the R_VEC macro to refer to vector types.
+ * 3. Call the functions from the API (described below), e.g. MyVector_init (&vec).
+ *    Note that these are auto-generated with the macro in step 1.
+ * 4. Do not access the fields in the vector struct directly, instead always use the functions from
+ *    the API.
+ *
+ * Vector API:
+ * - void R_VEC_FUNC(name, init)(R_VEC(name) *vec): Initializes an empty vector.
+ * - R_VEC(name) *R_VEC_FUNC(name, new)(): Allocates a new empty vector on the heap.
+ * - void R_VEC_FUNC(name, swap)(R_VEC(name) *vec_a, R_VEC(name) *vec_b): Swaps 2 vectors.
+ * - void R_VEC_FUNC(name, clear)(R_VEC(name) *vec, R_VEC_FINI(name) fini_fn, void *user):
+ *   Clears the vector by calling fini_fn for each element. The size is set to 0,
+ *   but the capacity remains the same (no allocation is freed).
+ * - void R_VEC_FUNC(name, fini)(R_VEC(name) *vec, R_VEC_FINI(name) fini_fn, void *user):
+ *   Clears the vector by calling fini_fn for each element. Also frees up all memory used
+ *   by the vector elements.
+ * - void R_VEC_FUNC(name, free)(R_VEC(name) *vec, R_VEC_FINI(name) fini_fn, void *user):
+ *   Similar to R_VEC_FUNC(name, fini), but also frees the vector itself.
+ * - size_t R_VEC_FUNC(name, length)(const R_VEC(name) *vec): Returns number of elements
+ *   in the vector.
+ * - size_t R_VEC_FUNC(name, capacity)(const R_VEC(name) *vec): Returns the max number of
+ *   elements that can fit in the vector before a re-allocation needs to be performed.
+ * - bool R_VEC_FUNC(name, empty)(const R_VEC(name) *vec): Returns a boolean value indicating
+ *   if the vector is empty or not.
+ * - type *R_VEC_FUNC(name, start_iter)(const R_VEC(name) *vec): Returns a pointer to the start
+ *   of the elements in the vector.
+ * - type *R_VEC_FUNC(name, end_iter)(const R_VEC(name) *vec): Returns a pointer to element PAST
+ *   the end of the last filled in element in the vector.
+ * - type *R_VEC_FUNC(name, at)(const R_VEC(name) *vec, size_t index): Returns a pointer to an
+ *   element in the vector. Note that this can be used for reading or writing from/to the element,
+ *   but not deleting (see the pop and remove functions for this).
+ * - R_VEC(name) *R_VEC_FUNC(name, clone)(const R_VEC(name) *vec): Creates a shallow clone of a vector.
+ * - void R_VEC_FUNC(name, reserve)(R_VEC(name) *vec, size_t new_capacity): Ensures the vector has
+ *   atleast a capacity of "new_capacity".
+ * - void R_VEC_FUNC(name, shrink_to_fit)(R_VEC(name) *vec): Shrinks the vector to exactly fit the
+ *   current number of elements it contains.
+ * - void R_VEC_FUNC(name, push_back)(R_VEC(name) *vec, type *value): Appends a single element to
```

Comment:
this is unnefficient. can we have a method to push_back by returning the address of the element to the caller, so it can save an extra memcpy?